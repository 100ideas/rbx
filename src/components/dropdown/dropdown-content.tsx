import classNames from "classnames";
import React from "react";

import { forwardRefAs, Generic, HelpersProps } from "../../base";

export type DropdownContentProps = HelpersProps;

export const DropdownContent = forwardRefAs<DropdownContentProps, "div">(
  ({ className, ...rest }, ref) => (
    <Generic
      className={classNames("dropdown-content", className)}
      ref={ref}
      {...rest}
    />
  ),
  { as: "div" },
);
