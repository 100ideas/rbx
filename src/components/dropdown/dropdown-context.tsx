import React from "react";

import { noop } from "../../utils";

export interface DropdownContextState {
  active: boolean;
  setActive: (value: boolean) => void;
}

export const initialState: DropdownContextState = {
  active: false,
  setActive: noop,
};

export const DropdownContext = React.createContext<DropdownContextState>(
  initialState,
);
