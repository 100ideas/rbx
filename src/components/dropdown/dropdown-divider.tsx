import classNames from "classnames";
import React from "react";

import { forwardRefAs, HelpersProps, transformHelpers } from "../../base";

export type DropdownDividerModifierProps = Partial<{ className: string }>;

export type DropdownDividerProps = HelpersProps & DropdownDividerModifierProps;

export const DropdownDivider = forwardRefAs<DropdownDividerProps, "hr">(
  (props, ref) => {
    const { as, ...rest } = transformHelpers(props);
    rest.className = classNames("dropdown-divider", rest.className);
    return React.createElement(as!, { ref, ...rest });
  },
  { as: "hr" },
);
