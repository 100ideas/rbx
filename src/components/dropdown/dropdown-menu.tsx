import classNames from "classnames";
import React from "react";

import { forwardRefAs, Generic } from "../../base";
import { HelpersProps } from "../../base/helpers";

export type DropdownMenuProps = HelpersProps;

export const DropdownMenu = forwardRefAs<DropdownMenuProps>(
  ({ className, ...rest }, ref) => (
    <Generic
      className={classNames("dropdown-menu", className)}
      ref={ref}
      {...rest}
    />
  ),
  { as: "div" },
);

DropdownMenu.displayName = "Dropdown.Menu";
