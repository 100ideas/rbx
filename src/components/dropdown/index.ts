import "./dropdown.module.sass";

export { Dropdown } from "./dropdown";
