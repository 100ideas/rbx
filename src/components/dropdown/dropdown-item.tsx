import { cx } from "emotion";
import React from "react";

import { forwardRefAs } from "@/base";
import { ModifierProps, transformModifiers } from "@/modifiers";
import { DropdownContext } from "./dropdown-context";

export type DropdownItemModifierProps = Partial<{
  active: boolean;
}>;

export type DropdownItemProps = ModifierProps & DropdownItemModifierProps;

export const DropdownItem = forwardRefAs<DropdownItemProps, "a">(
  (props, ref) => {
    const { as, active, onClick, ...rest } = transformModifiers(props);
    rest.className = cx("dropdown-item", rest.className, {
      "is-active": active,
    });
    return (
      <DropdownContext.Consumer>
        {ctx =>
          React.createElement(as!, {
            onClick: event => {
              if (onClick) {
                onClick(event);
              }
              ctx.setActive(!ctx.active);
            },
            ref,
            ...rest,
          })
        }
      </DropdownContext.Consumer>
    );
  },
  "a",
);
