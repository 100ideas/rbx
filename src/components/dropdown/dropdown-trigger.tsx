import classNames from "classnames";
import PropTypes from "prop-types";
import React from "react";

import { forwardRefAs, Generic, HelpersProps } from "../../base";
import { DropdownContext, DropdownContextValue } from "./dropdown-context";

export type DropdownTriggerModifierProps = Partial<{
  onClick: React.MouseEventHandler;
}>;

export type DropdownTriggerProps = HelpersProps & DropdownTriggerModifierProps;

const onClickHandler = (
  onClick: DropdownTriggerProps["onClick"] | undefined,
  ctx: DropdownContextValue,
) => (event: React.MouseEvent) => {
  if (onClick !== undefined) {
    onClick(event);
  }
  ctx.setActive(!ctx.active);
};

export const DropdownTrigger = forwardRefAs<
  HTMLDivElement,
  DropdownTriggerProps
>(
  ({ className, onClick, ...rest }, ref) => (
    <DropdownContext.Consumer>
      {ctx => (
        <Generic
          className={classNames("dropdown-trigger", className)}
          onClick={onClickHandler(onClick, ctx)}
          ref={ref}
          {...rest}
        />
      )}
    </DropdownContext.Consumer>
  ),
  { as: "div" },
);

DropdownTrigger.displayName = "Dropdown.Trigger";
DropdownTrigger.propTypes = {
  onClick: PropTypes.func,
};
