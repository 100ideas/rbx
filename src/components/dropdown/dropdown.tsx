import { cx } from "emotion";
import React, { PureComponent } from "react";

import { Button } from "components/button";
import { Icon } from "components/icon";
import { classNames, clean, ModifierProps } from "modifiers";
import { Colors } from "modifiers/colors";
import { DropdownDivider } from "./dropdown-divider";
import { DropdownItem } from "./dropdown-item";

export type DropdownModifierProps = Partial<{
  align: "right";
  children: React.ReactNode;
  className: string;
  color: Colors;
  hoverable: boolean;
  onChange: (value: string) => void;
  ref: React.RefObject<HTMLDivElement>;
  style: React.CSSProperties;
  value: any;
}>;

export type DropdownProps = ModifierProps &
  DropdownModifierProps &
  Partial<
    Omit<
      React.ComponentPropsWithoutRef<"div">,
      "color" | "onChange" | "unselectable"
    >
  >;

const initialState = {
  open: false,
};

export type DropdownState = typeof initialState;

export class Dropdown extends PureComponent<DropdownProps, DropdownState> {
  public static Item = DropdownItem;
  public static Divider = DropdownDivider;

  public static defaultProps = {
    children: [],
  };

  public readonly state: DropdownState = initialState;

  private htmlElement: React.RefObject<HTMLDivElement>;
  private listener: (() => void) | null = null;

  constructor(props: DropdownProps) {
    super(props);
    this.htmlElement = props.ref
      ? props.ref
      : React.createRef<HTMLDivElement>();
  }

  public componentDidMount() {
    this.listener = () => this.close;
    document.addEventListener("click", this.listener);
  }

  public componentWillUnmount() {
    document.removeEventListener("click", this.listener!);
    this.listener = null;
  }

  public toggle = (evt: React.MouseEvent<HTMLDivElement>) => {
    if (this.props.hoverable) {
      return;
    }
    if (evt) {
      evt.preventDefault();
    }
    this.setState(({ open }) => ({ open: !open }));
  }

  public select = (value: string) => () => {
    if (this.props.onChange) {
      this.props.onChange(value);
    }
    this.close();
  }

  public render() {
    const {
      className,
      children,
      value,
      color,
      align,
      hoverable,
      onChange,
      ...allProps
    } = this.props;

    let current = null;
    const props = clean(allProps);

    const childrenArray = React.Children.map(children, (child, i) => {
      if (typeof child !== "string" && typeof child !== "number") {
        if (i === 0 || child.props.value === value) {
          current = child.props.children;
        }
        return React.cloneElement(
          child,
          child.type === DropdownItem
            ? {
                active: child.props.value === value,
                onClick: this.select(child.props.value),
              }
            : {},
        );
      }
      return child;
    });

    return (
      <div
        {...props}
        ref={this.htmlElement}
        className={cx("dropdown", classNames(allProps), className, {
          "is-active": this.state.open,
          [`is-,${align}`]: align,
          "is-hoverable": hoverable,
        })}
      >
        <div
          className="dropdown-trigger"
          role="presentation"
          onClick={this.toggle}
        >
          <Button color={color}>
            <span>{current}</span>
            <Icon icon="angle-down" size="small" />
          </Button>
        </div>
        <div className="dropdown-menu" id="dropdown-menu" role="menu">
          <div className="dropdown-content">{childrenArray}</div>
        </div>
      </div>
    );
  }

  private close(event?: MouseEvent) {
    // IDK yet how to test using the ref in enzyme
    if (
      this.props.hoverable ||
      (event &&
        this.htmlElement.current &&
        event.target instanceof HTMLElement &&
        !this.htmlElement.current.contains(event.target))
    ) {
      return;
    }
    this.setState({ open: false });
  }
}
