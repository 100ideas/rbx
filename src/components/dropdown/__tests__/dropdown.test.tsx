import Enzyme from "enzyme";
import React from "react";

import { Dropdown, DROPDOWN_ALIGNMENTS, DropdownController } from "../dropdown";
import { DropdownContent } from "../dropdown-content";
import { DropdownContext } from "../dropdown-context";
import { DropdownDivider } from "../dropdown-divider";
import { DropdownItem } from "../dropdown-item";
import { DropdownMenu } from "../dropdown-menu";
import { DropdownTrigger } from "../dropdown-trigger";

import { hasProperties } from "@/__tests__/helpers";

describe("Card component", () => {
  hasProperties(Dropdown, {
    Content: DropdownContent,
    Context: DropdownContext,
    Divider: DropdownDivider,
    Item: DropdownItem,
    Menu: DropdownMenu,
    Trigger: DropdownTrigger,
    defaultProps: { as: "div" },
  });

  it("should render as the default component", () => {
    const wrapper = Enzyme.mount(<Dropdown />);
    try {
      expect(
        wrapper
          .find(DropdownController)
          .children()
          .is("div"),
      ).toBe(true);
    } finally {
      wrapper.unmount();
    }
  });

  it("should render as a custom component", () => {
    const as = "span";
    const wrapper = Enzyme.mount(<Dropdown as={as} />);
    try {
      expect(
        wrapper
          .find(DropdownController)
          .children()
          .is(as),
      ).toBe(true);
    } finally {
      wrapper.unmount();
    }
  });

  it("should forward ref", () => {
    const ref = React.createRef<HTMLDivElement>();
    // Enzyme owns outer ref: https://github.com/airbnb/enzyme/issues/1852
    const wrapper = Enzyme.mount(
      <div>
        <Dropdown ref={ref} />
      </div>,
    );
    try {
      expect(ref.current).toBe(wrapper.find(".dropdown").instance());
    } finally {
      wrapper.unmount();
    }
  });

  it("should have bulma className", () => {
    const wrapper = Enzyme.mount(<Dropdown />);
    try {
      expect(
        wrapper
          .find(DropdownController)
          .children()
          .hasClass("dropdown"),
      ).toBe(true);
    } finally {
      wrapper.unmount();
    }
  });

  it("should preserve custom className", () => {
    const className = "foo";
    const wrapper = Enzyme.mount(<Dropdown className={className} />);
    try {
      expect(
        wrapper
          .find(DropdownController)
          .children()
          .hasClass(className),
      ).toBe(true);
    } finally {
      wrapper.unmount();
    }
  });

  it("should be active", () => {
    const wrapper = Enzyme.mount(<Dropdown active />);
    try {
      expect(
        wrapper
          .find(DropdownController)
          .children()
          .hasClass("is-active"),
      ).toBe(true);
    } finally {
      wrapper.unmount();
    }
  });

  it("should be hoverable", () => {
    const wrapper = Enzyme.mount(<Dropdown hoverable />);
    try {
      expect(
        wrapper
          .find(DropdownController)
          .children()
          .hasClass("is-hoverable"),
      ).toBe(true);
    } finally {
      wrapper.unmount();
    }
  });

  it("should be up", () => {
    const wrapper = Enzyme.mount(<Dropdown up />);
    try {
      expect(
        wrapper
          .find(DropdownController)
          .children()
          .hasClass("is-up"),
      ).toBe(true);
    } finally {
      wrapper.unmount();
    }
  });

  DROPDOWN_ALIGNMENTS.map(align =>
    it(`should be aligned ${align}`, () => {
      const wrapper = Enzyme.mount(<Dropdown align={align} />);
      try {
        expect(
          wrapper
            .find(DropdownController)
            .children()
            .hasClass(`is-${align}`),
        ).toBe(true);
      } finally {
        wrapper.unmount();
      }
    }),
  );

  [false, true].map(managed =>
    it(`should ${
      managed ? "not " : ""
    }be disabled on click in document when managed is ${managed}`, () => {
      const wrapper = Enzyme.mount(<Dropdown active managed={managed} />);
      try {
        document.getElementsByTagName("body")[0].click();
        wrapper.update();
        expect(wrapper.find(".dropdown").hasClass("is-active")).toBe(managed);
      } finally {
        wrapper.unmount();
      }
    }),
  );
});
