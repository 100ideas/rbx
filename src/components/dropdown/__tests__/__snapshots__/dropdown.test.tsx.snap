// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component Should Exist 1`] = `[Function]`;

exports[`Dropdown component Should concat Bulma class with classes in props 1`] = `
<div
  className="dropdown other-class"
>
  <div
    className="dropdown-trigger"
    onClick={[Function]}
    role="presentation"
  >
    <button
      className="button"
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
    >
      <span>
        Item
      </span>
      <span
        className="icon is-small"
      >
        <i
          className="rbx rbx-angle-down"
        />
      </span>
    </button>
  </div>
  <div
    className="dropdown-menu"
    id="dropdown-menu"
    role="menu"
  >
    <div
      className="dropdown-content"
    >
      <div
        className="dropdown-item is-active"
        onClick={[Function]}
        role="presentation"
        title="value"
      >
        Item
      </div>
    </div>
  </div>
</div>
`;

exports[`Dropdown component Should have custom inline styles 1`] = `
<div
  className="dropdown"
  style={
    Object {
      "width": 400,
    }
  }
>
  <div
    className="dropdown-trigger"
    onClick={[Function]}
    role="presentation"
  >
    <button
      className="button"
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
    >
      <span>
        Item
      </span>
      <span
        className="icon is-small"
      >
        <i
          className="rbx rbx-angle-down"
        />
      </span>
    </button>
  </div>
  <div
    className="dropdown-menu"
    id="dropdown-menu"
    role="menu"
  >
    <div
      className="dropdown-content"
    >
      <div
        className="dropdown-item is-active"
        onClick={[Function]}
        role="presentation"
        title="value"
      >
        Item
      </div>
    </div>
  </div>
</div>
`;

exports[`Dropdown component Should have divider 1`] = `
<div
  className="dropdown"
  style={
    Object {
      "width": 400,
    }
  }
>
  <div
    className="dropdown-trigger"
    onClick={[Function]}
    role="presentation"
  >
    <button
      className="button"
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
    >
      <span>
        Item
      </span>
      <span
        className="icon is-small"
      >
        <i
          className="rbx rbx-angle-down"
        />
      </span>
    </button>
  </div>
  <div
    className="dropdown-menu"
    id="dropdown-menu"
    role="menu"
  >
    <div
      className="dropdown-content"
    >
      <div
        className="dropdown-item is-active"
        onClick={[Function]}
        role="presentation"
        title="value"
      >
        Item
      </div>
      <hr
        className="dropdown-divider"
      />
      <div
        className="dropdown-item"
        onClick={[Function]}
        role="presentation"
        title="other"
      >
        Other
      </div>
    </div>
  </div>
</div>
`;

exports[`Dropdown component Should have dropdown classname 1`] = `
<div
  className="dropdown"
>
  <div
    className="dropdown-trigger"
    onClick={[Function]}
    role="presentation"
  >
    <button
      className="button"
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
    >
      <span />
      <span
        className="icon is-small"
      >
        <i
          className="rbx rbx-angle-down"
        />
      </span>
    </button>
  </div>
  <div
    className="dropdown-menu"
    id="dropdown-menu"
    role="menu"
  >
    <div
      className="dropdown-content"
    />
  </div>
</div>
`;
